packages/pyright-internal/src/analyzer/dataClasses.ts

```diff
@@ -654,6 +673,93 @@ export function synthesizeDataClassMethods(
     );
 }
 
+// Validates converter and, if valid, returns its input type. If invalid,
+// fieldType is returned.
+function getConverterInputType(
+    evaluator: TypeEvaluator,
+    converterNode: ArgumentNode,
+    fieldType: Type,
+    fieldName: string
+): Type {
+    const converterType = evaluator.getTypeOfExpression(converterNode.valueExpression).type;
+
+    if (!isFunction(converterType) && !isOverloadedFunction(converterType)) {
+        return fieldType;
```

Comment:
Do you want to emit a diagnostic in this case rather than simply bailing? I could argue both sides of that.

I mentioned previous that it's probably OK to ignore unions for now. One other case that might be more common is a callable object. That would also be the same case as a object whose type conforms to a callback protocol. It may be fine to ignore for now and see if it ever arises.

Reply:
Ah, I had naively thought that the `argAssignmentParam` would be reported in scenarios that I wasn't supporting.

Can you give me a code sample showing a union of functions/overloads? I wasn't able to figure out what syntax you were thinking of.

Here's what I found while playing with `Callable` and `Protocol` objects:

### Callable class instances
Instances of objects with `__call__` methods, as in the example below, don't work currently. Accepting `converterType` values that are class instances and validating them the same way I'm validating functions would help, but I'd also need to check if `__call__` is overloaded and special case that in the `OverloadedFunction` code path.

Is there some more general way to handle this? I saw `isCallableType` and thought that might help, but I'd still need a way to determine if there were any overloads in play and obtain them. Maybe some additional helper functions would be useful -- `isOverloadedCallableType` / `getOverloadsOfCallableType`?

Btw, I also noticed that if you remove the assignment of `callback`, the `TypeCategory` of the `converter` arg expression becomes `Unbound` and we fall back to returning `fieldType`. However, we also show the `reportUnboundVariable` diagnostic in that case, so maybe that's sufficient.

```python
class Callback(Protocol):
    def __call__(self, arg1: str) -> int: ...

class CallableObject(Callback):
    def __call__(self, arg1: str) -> int:
        return 1

callback: Callback = CallableObject()

@dataclass
class Foo:
    field5: int = field(converter=callback)
```

### Function as a Protocol
If I assign a function to the `Protocol` variable instead, validation of params, return value, etc work because we know that it is a function.

```python
class Callback(Protocol):
    def __call__(self, arg1: str) -> int: ...

def function1(arg1: str) -> int:
    return 1

callback: Callback = function1 

@dataclass
class Foo:
    field5: int = field(converter=callback)
```

### Callable
Similarly, validation works in the following `Callable` scenario, although if `callable` is unassigned we don't show the `reportUnboundVariable` diagnostic as we do for the `Protocol` case above, but instead get `object` from the `applySolvedTypeVars` call and use that as the `effectiveType` of the field. 

Is the difference in behavior with these two "unassigned" scenarios expected? Should an unassigned `Callable` also have `TypeCategory.Unbound`?

```python
def function1(arg1: str) -> int:
    return 1

callable: Callable[[str], int] = function1

@dataclass
class Foo:
    field5: int = field(converter=callable)
```

Reply:
Here's an (admittedly contrived) example of a union type:

```python
def get_class(converter: Callable[[str], str] | Callable[[int], str]):
    @dataclass
    class Foo:
        field5: int = field(converter=converter)

    return Foo
```

Here's another variant:

```python
support_str = load_config()
def fn1(val: int) -> int: ...
def fn2(val: str) -> int: ...

@dataclass
class Foo:
    field5: int = field(converter=fn2 if support_str else fn1)
```

Like I said, it's probably fine to ignore unions for now.

A an object with a `__call__` method internally is a `ClassType`, so your current logic will reject it. In particular, it's an instantiated class (i.e. `isClassInstance` will return true). You can call `getBoundMethod(type, '__call__')` to determine whether an instantiated class has a `__call__` method, and if so what its signature is. This will return a `FunctionType`, `OverloadedFunctionType` or `undefined` depending on the type and presence of a `__call__` method.